README.md
# Bitmap Transformer API lab for CodeFellows 401

### Collaborators
 - Mugsy Carter
 - Tim Combs
 - Nathan Hugon

### Project Functionality
  - This is a Code Fellows Lab Assignment to produce a Bitmap Transformer
  - It takes in a Windows BMP format file and produces a value invert of that file
  - In the codebase is also a commented out grayscale transformation functionality

### How To Use Codebase
  - This module uses Node and the following modules:
    - eslint, mocha and rimraf for testing
  - Please refer to the package.json for more info
    - chai IS listed as devDependency for future development

### Use Cases
  - read.readWrite is the method that is exposed via require('./lib/read')
    - it takes a file path to a bmp file as a parameter
    - it writes a value inverted bmp, another-file.bmp to the root directory of the project

### Code Shape
  - This code has been vetted using Eslint and was reviewed by Code Fellows using Travis-CI

### Collborations/Questions/issues
  - Not currently looking for collaborators at this time
  - Any questions and concerns can be handled by opening an issue on the codebase

### License
  - Licensed under the MIT license - see LICENSE.md for more info


CHANGELOG.md
# Change Log
All changes to this project will be documented in this CHANGELOG

Format is based on Keep a Changelong (http://keepachangelog.com/) and the project adheres to Semantic Versioning (http://semver.org).


## [1.1.0] - 2016-10-22
### Changed
- Refactored tests to fat arrow notation

### Added
- Created gray scale transformation functionality
- README.md, CHANGELOG.md

## [1.0.0] - 2016-10-21
### Changed
- First release with MVP functionality

### Added
- Add value invert functionality
- Tests working properly

test/test.js
const read = require('../lib/read');
const assert = require('assert');
const fs = require('fs');
const rimraf = require('rimraf');

describe('testing bitmap transfomer', () => {
  beforeEach(function(done) {
    rimraf('./another-file.bmp', done);
    console.log('rimraffed');
  });

  it('adds a new altered image', (done) => {
    console.log('adding image test running');
    read.readWrite('./non-palette-bitmap.bmp', () => {
      assert.ok(fs.existsSync('./another-file.bmp'));
      done();
    });
  });

  it('inverts the colrs, byte by byte', (done) => {
    console.log('value inversion test running');
    read.readWrite('./non-palette-bitmap.bmp', (data) => {
      //know specific byte info at buffer index 25000 = 132
      //color channel 0 - 255, so subtract 123 from 255
      assert.deepEqual(132, data[25000]);
      done();
    });
  });

  it('returns the proper transform', (done) => {
    console.log('proper transform test running');
    //run the transformation
    read.readWrite('non-palette-bitmap.bmp', (data) => {
      //get the golden chicken buffer
      fs.readFile('gc-value-invert.bmp', (err, gcBuf) => {
        if (err) throw err;
        //compare golden chicken vs transformed file
        assert.ok(gcBuf.equals(data));
        done();
      });
    });
  });

});

notes.js





//In computing, although the grayscale can be computed through rational numbers, image pixels are stored in binary, quantized form. Some early grayscale monitors can only show up to sixteen (4-bit) different shades, but today grayscale images (as photographs) intended for visual display (both on screen and printed) are commonly stored with 8 bits per sampled pixel, which allows 256 different intensities (i.e., shades of gray) to be recorded, typically on a non-linear scale. The precision provided by this format is barely sufficient to avoid visible banding artifacts, but very convenient for programming because a single pixel then occupies a single byte.


const fs = require('fs');
var read = {};

read.readWrite = function readWrite(cb) {
  fs.readFile('./non-palette-bitmap.bmp', (err, data) => {
    if (err) throw err;
    console.log(data);

    //define new buffer and copy orig buffer to it
    const bufSize = data.length;
    const newBuf = Buffer.alloc(bufSize);
    data.copy(newBuf);

    //get image body offset from specific place in file
    const imgOffset = data.readInt32LE(10);
    console.log('imgOffset', imgOffset);
    //get bitsPerPixel
    const bitsPerPixel = data.readInt16LE(28);
    console.log('bpp', bitsPerPixel);

    //value invert
    for (let i = imgOffset; i < bufSize; i++) {
      newBuf[i] = 255 - newBuf[i];
    }

    for (let i = imgOffset; i < bufSize - 2; i + 3) {
      newBuf[i] = hhhh;
      //0.21 * r + 0.72 * g + 0.07 * b
      //0.21 * newBuf[i] + 0.72 * newBuf[i + 1] + 0.07 * newBuf[i + 2]
    }

    fs.writeFile('./another-file.bmp', newBuf, cb);

  });
};

module.exports = read;

if (i % 3 === 0 || i === 0) {
        var rTransform = (0.21 * parseInt(newBuf[i], 16)).toString(16);
        var gTransform = (0.72 * parseInt(newBuf[i], 16)).toString(16);
        var bTransform = (0.07 * parseInt(newBuf[i], 16)).toString(16);
        
      }
    }

        // //broken grayscale
        // console.log('r', rTransform, 'g', gTransform, 'b', bTransform);
        // newBuf[i] = rTransform;
        // newBuf[i + 1] = gTransform;
        // newBuf[i + 2] = bTransform;
        // //0.21 * r + 0.72 * g + 0.07 * b
        // //0.21 * newBuf[i] + 0.72 * newBuf[i + 1] + 0.07 * newBuf[i + 2]

    //working value invert
    // const rLum = 0.21;
    // const gLum = 0.72;



    // const bLum = 0.07;
    // console.log('r', rLum.toString(16), 'g', gLum.toString(16), 'b', bLum.toString(16));
    // newBuf[i] = rLum.toString(16) * newBuf[i];
    // newBuf[i + 1] = gLum.toString(16) * newBuf[i + 1];
    // newBuf[i + 2] = bLum.toString(16) * newBuf[i + 2];


    //framework if i was going to use streams
    // var readStream = fs.createReadStream('./non-palette-bitmap.bmp');
    // var writeStream = fs.createWriteStream('./alt-bitmap.bmp');

    // readStream   // reads from myfile.txt
    //   .pipe(transformImage) //need to write this function
    //   .pipe(writeStream)  // writes to myfile.encrypted
    //           //eventEmitter & eventListener
    //   .on('finishTransform', (err) => {
    //     if (err) throw err;
    //     console.log('done encrypting');
    //   });


lib/read.js
const fs = require('fs');
var read = {};

read.readWrite = function readWrite(filepath, cb) {
  fs.readFile(filepath, (err, data) => {
    if (err) throw err;

    //define new buffer and copy orig buffer to it
    const bufSize = data.length;
    const newBuf = Buffer.alloc(bufSize);
    data.copy(newBuf);
    
    //to get pixel to test before mutation
    //console.log('data', data[25000]);

    //object to hold header and file info as properties
    read.fileData = {
      fileBuf: newBuf,
      fileHeader: newBuf.readIntLE(0,2).toString(16),
      fileSize: newBuf.readIntLE(2,6).toString(16),
      resArea1: newBuf.readIntLE(6,8).toString(16),
      resArea2: newBuf.readIntLE(8,10).toString(16),
      imgOffset: newBuf.readInt32LE(10),
      bitsPerPixel: data.readInt16LE(28)
    };

      //working visual invert
    for (let i = read.fileData.imgOffset; i < bufSize; i++) {
      read.fileData.fileBuf[i] = 255 - read.fileData.fileBuf[i];
    }

    //
    // //operates on three bytes at once
    // for (let i = read.fileData.imgOffset; i < bufSize - 2; i++) {
    //   if (i % 3 === 0) {
    //     //working gray scale by avg
    //     var pixelAvg = (read.fileData.fileBuf[i] + read.fileData.fileBuf[i+1] + read.fileData.fileBuf[i+2])/3;
    //     for (let j = 0; j < 3; j++) {
    //       newBuf[i+j] = pixelAvg;
    //     }
    //   }
    // }
    
    fs.writeFile('./another-file.bmp', newBuf, (err) => {
      if (err) return err;
      cb(newBuf);
      return newBuf;
    });
  });
};

module.exports = read;

    //potential framework leading toward using streams
    // var readStream = fs.createReadStream('./non-palette-bitmap.bmp');
    // var writeStream = fs.createWriteStream('./alt-bitmap.bmp');

    // readStream   // reads from myfile.txt
    //   .pipe(transformImage) //need to write this function
    //   .pipe(writeStream)  // writes to myfile.encrypted
    //           //eventEmitter & eventListener
    //   .on('finishTransform', (err) => {
    //     if (err) throw err;
    //     console.log('done encrypting');
    //   });
